// @takeshape/ssg/lib/graphql/ast.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.parse = parse;
exports.createArgument = createArgument;
exports.createVariableDefinition = createVariableDefinition;
exports.createSelectionSet = createSelectionSet;
exports.createField = createField;
exports.isQueryNode = isQueryNode;
exports.findQuery = findQuery;
exports.addQueryVariables = addQueryVariables;
exports.wrapSelectionSet = wrapSelectionSet;
exports.wrapQuery = wrapQuery;

var _graphql = require("graphql");

function parse(source, options) {
  return (0, _graphql.parse)(source, options);
}

function createArgument({
  name,
  value
}) {
  return {
    kind: 'Argument',
    name: {
      kind: 'Name',
      value: name
    },
    value: {
      kind: 'Variable',
      name: {
        kind: 'Name',
        value
      }
    }
  };
}

function notNull(type) {
  return {
    kind: 'NonNullType',
    type
  };
}

function createType(typeString) {
  const isNotNull = typeString.endsWith('!');
  const value = isNotNull ? typeString.slice(0, typeString.length - 1) : typeString;
  const namedType = {
    kind: 'NamedType',
    name: {
      kind: 'Name',
      value
    }
  };
  return isNotNull ? notNull(namedType) : namedType;
}

function createVariableDefinition({
  name,
  type
}) {
  return {
    kind: 'VariableDefinition',
    directives: [],
    variable: {
      kind: 'Variable',
      name: {
        kind: 'Name',
        value: name
      }
    },
    type: createType(type)
  };
}

function createSelectionSet(selections) {
  return {
    kind: 'SelectionSet',
    selections
  };
}

function createField(name) {
  return {
    kind: 'Field',
    name: {
      kind: 'Name',
      value: name
    },
    arguments: [],
    directives: []
  };
}

function isQueryNode(node) {
  return node.kind === 'OperationDefinition' && node.operation === 'query';
}

function findQuery(ast) {
  return ast.definitions.find(isQueryNode);
}

function addQueryVariables(variables, ast) {
  const query = findQuery(ast);

  if (query) {
    const variableNames = new Set(variables.map(variable => variable.name));

    if (query.variableDefinitions) {
      query.variableDefinitions = query.variableDefinitions.filter(def => !variableNames.has(def.variable.name.value)).concat(variables.map(createVariableDefinition));
    }
  }

  return ast;
}

function wrapSelectionSet(selectionSet, name, args = []) {
  return {
    kind: 'SelectionSet',
    selections: [{
      kind: 'Field',
      name: {
        kind: 'Name',
        value: name
      },
      arguments: args,
      directives: [],
      selectionSet
    }]
  };
}

function wrapQuery(name, args, ast) {
  const query = findQuery(ast);

  if (query) {
    const argsAst = args.map(createArgument);
    query.selectionSet = wrapSelectionSet(query.selectionSet, name, argsAst);
  }

  return ast;
}
