// @takeshape/ssg/lib/graphql/schema-connector-factory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getRemoteConnector = getRemoteConnector;
exports.schemaConnectorFactory = schemaConnectorFactory;

var _fsExtra = _interopRequireDefault(require("fs-extra"));

var _nodeFetch = _interopRequireDefault(require("node-fetch"));

var _path = _interopRequireDefault(require("path"));

var _graphql = require("graphql");

var _migrate = require("./migrate");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getRemoteConnector(config) {
  const {
    url
  } = config;

  if (!url) {
    throw new Error('Invalid GraphQL config: missing url');
  }

  return params => {
    return (0, _nodeFetch.default)(url, {
      method: 'POST',
      body: JSON.stringify(params),
      headers: {
        'Content-Type': 'application/json',
        ...(config.headers || {})
      }
    }).then(res => {
      if (res.ok) {
        return res.json();
      }

      return res.text().then(text => {
        throw new Error(`${res.statusText}: ${text}`);
      });
    });
  };
}

function parseConfigJson(json) {
  if (json.request) {
    return {
      type: 'request',
      ...json.request
    };
  }

  if (json['graphql-js']) {
    return {
      type: 'graphql-js',
      file: json['graphql-js']
    };
  }

  return null;
}

function readConfigFile(path) {
  let json;

  try {
    json = JSON.parse(_fsExtra.default.readFileSync(`${path}/.graphqlrc`, 'utf-8'));
  } catch (e) {
    throw new Error(`An error occurred loading .graphqlrc.`);
  }

  const config = parseConfigJson(json);

  if (config) {
    return config;
  }

  throw new Error(`Invalid configuration file: ${JSON.stringify(json)}`);
}

function schemaConnectorFactory(cwd, config) {
  config = config || readConfigFile(cwd);
  let connector;

  if (config.type === 'graphql-js' && config.file) {
    const schema = require(_path.default.join(cwd, config.file)); // eslint-disable-line


    connector = async params => {
      const res = await (0, _graphql.graphql)(schema.default || schema, params.query, null, {}, params.variables);
      return {
        data: res.data,
        errors: res.errors && res.errors.map(_graphql.formatError)
      };
    };
  }

  if (config.type === 'request') {
    connector = getRemoteConnector(config);
  }

  if (connector) {
    return (0, _migrate.connectorWithMigrations)(connector);
  }

  throw new Error('Could not create GraphQL connector - invalid configuration');
}
