// @takeshape/cli/lib/auth.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.verifyAuth = verifyAuth;
exports.parseToken = parseToken;
exports.login = login;
exports.projectAuth = projectAuth;

var _nodeFetch = _interopRequireDefault(require("node-fetch"));

var _jsonwebtoken = _interopRequireDefault(require("jsonwebtoken"));

var _errors = require("./errors");

var _api = require("./util/api");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function verifyAuth(authToken) {
  if (!authToken) {
    return {
      isValid: false,
      error: 'missing'
    };
  }

  if ((0, _api.isJWT)(authToken)) {
    const authJwt = _jsonwebtoken.default.decode(authToken);

    const now = Date.now() / 1000;

    if (!authJwt || typeof authJwt !== 'object' || now > authJwt.exp) {
      return {
        isValid: false,
        error: 'invalid'
      };
    }
  }

  return {
    isValid: true,
    error: null
  };
}

async function parseToken(res) {
  try {
    const body = await res.json();

    if (body.token) {
      return body.token;
    }
  } catch (e) {// ignore parsing error
  }

  const cookie = res.headers.get('set-cookie');

  if (cookie) {
    const matches = /s=(.+?);/.exec(cookie);

    if (matches) {
      return matches[1];
    }
  }

  throw new _errors.HTTPError('Forbidden', 403);
}

async function fetchToken(method, endpoint, body, headers) {
  const params = {
    method,
    headers: {
      'Content-Type': 'application/json',
      ...headers
    }
  };

  if (body) {
    params.body = JSON.stringify(body);
  }

  const res = await (0, _nodeFetch.default)(endpoint, params);

  if (res.ok) {
    return parseToken(res);
  }

  throw new _errors.HTTPError(`${res.statusText}`, res.status);
}

function login(endpoint, email, password) {
  return fetchToken('POST', `${endpoint}/login`, {
    email,
    password
  });
}

async function projectAuth(endpoint, projectId, authToken) {
  try {
    return await fetchToken('GET', `${endpoint}/project/${projectId}/access-token`, null, (0, _api.getAuthHeader)(authToken));
  } catch (e) {
    throw (0, _errors.formatErrorMessage)(e, 'getting developer access token');
  }
}
