// @takeshape/cli/lib/graphql.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getGraphQLEndpoint = getGraphQLEndpoint;
exports.createConnector = createConnector;
exports.graphQLQuery = graphQLQuery;

var _nodeFetch = _interopRequireDefault(require("node-fetch"));

var _asyncRetry = _interopRequireDefault(require("async-retry"));

var _errors = require("./errors");

var _api = require("./util/api");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getGraphQLEndpoint(params, projectId) {
  const project = projectId ? `project/${projectId}/` : '';
  return `${params.endpoint}/${project}graphql`;
}

const defaultOptions = {
  retries: 3,
  timeout: 0
};

function createConnector(params, authToken, projectId, options = defaultOptions) {
  const endpoint = getGraphQLEndpoint(params, projectId);
  const authHeader = (0, _api.getAuthHeader)(authToken);
  return async params => (0, _asyncRetry.default)(async bail => {
    const res = await (0, _nodeFetch.default)(endpoint, {
      method: 'POST',
      timeout: options.timeout,
      body: JSON.stringify(params),
      headers: {
        'Content-Type': 'application/json',
        ...authHeader
      }
    });
    const body = await res.json();

    if (res.ok) {
      return body;
    }

    const error = new _errors.HTTPError(`${res.statusText}: ${body.message}`, res.status);

    if (res.status === 401 || res.status === 403) {
      // bail to stop retrying
      return bail(error);
    }

    throw error;
  }, {
    retries: options.retries
  });
}

function graphQLQuery(query) {
  return async (client, variables) => {
    const {
      data,
      errors
    } = await client({
      query,
      variables
    });
    if (errors) throw new Error(errors.map(error => error.message).join(', '));
    return data.result;
  };
}
