// @takeshape/cli/lib/saga.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.call = call;
exports.all = all;
exports.isGenerator = isGenerator;
exports.runSaga = runSaga;

function call(fn, ...args) {
  return {
    type: 'CALL',
    fn,
    args
  };
}

function all(effects) {
  return {
    type: 'ALL',
    effects
  };
}

function isGenerator(obj) {
  return obj && typeof obj.next === 'function' && typeof obj.throw === 'function';
}

const handlers = {
  async CALL(effect) {
    const value = effect.fn.apply(null, effect.args);

    if (isGenerator(value)) {
      return runSaga(value); // eslint-disable-line @typescript-eslint/no-use-before-define
    }

    return value;
  },

  ALL(effect) {
    return Promise.all(effect.effects.map(childEffect => {
      if (handlers[childEffect.type]) {
        return handlers[childEffect.type](childEffect);
      }

      throw new Error('Unhandled Effect');
    }));
  }

};

async function runSaga(saga, lastInput, isError = false) {
  const {
    value,
    done
  } = saga[isError ? 'throw' : 'next'](lastInput);

  if (done) {
    return value;
  }

  if (handlers[value.type]) {
    try {
      const input = await handlers[value.type](value);
      return runSaga(saga, input, false);
    } catch (e) {
      return runSaga(saga, e, true);
    }
  }

  throw new Error('Unhandled Effect');
}
