// @takeshape/ssg/es/config.js
import yaml from 'js-yaml';
import jv from 'json-variables';
import flatten from 'lodash/flatten';
import forEach from 'lodash/forEach';
import merge from 'lodash/merge';
export function prepareConfig(config, envConfig) {
  const combined = merge({
    env: {},
    context: {},
    routes: {},
    imageBaseUrl: 'https://images.takeshape.io',
    assetBaseUrl: 'https://assets.takeshape.io',
    imageDefaults: {
      auto: 'compress,format'
    },
    locale: 'en-us',
    pathPrefix: '',
    templatePath: 'templates',
    buildPath: 'build',
    staticPath: 'static',
    dates: {
      tz: 'America/New_York',
      format: 'LLL'
    },
    usageStats: false,
    gzip: false
  }, config, envConfig);
  return jv(combined, {
    heads: '${',
    tails: '}'
  });
}
export async function loadConfig(fileLoader, configFilePath, envConfig) {
  const config = yaml.load((await fileLoader(configFilePath || 'tsg.yml')));
  return prepareConfig(config, envConfig);
}
export function localizeConfig(config, localeStr) {
  if (!config.locales || !config.locales[localeStr]) {
    throw new Error(`Locale "${localeStr}" is not configured `);
  }

  const locale = config.locales[localeStr];
  return { ...config,
    pathPrefix: locale.pathPrefix,
    locale: localeStr,
    dates: { ...config.dates,
      ...locale.dates
    },
    numbers: { ...config.numbers,
      ...locale.numbers
    }
  };
}

function isGraphQLPath(path) {
  return path.endsWith('.graphql') || path.endsWith('.gql');
}

export function isGraphQLQueryConfig(x) {
  return typeof x === 'object' && typeof x.query === 'string' && isGraphQLPath(x.query);
}

function getContextGraphQLQueries(context, info) {
  if (typeof context === 'string' && isGraphQLPath(context)) {
    return [{
      filePath: context,
      ...info
    }];
  }

  if (typeof context === 'object') {
    if (isGraphQLQueryConfig(context)) {
      return [{
        filePath: context.query,
        ...info
      }];
    }

    const result = [];
    forEach(context, value => {
      Array.prototype.push.apply(result, getContextGraphQLQueries(value, info));
    });
    return result;
  }

  return [];
}

export function getGraphQLQueries(config) {
  const {
    context,
    routes
  } = config;
  const routeQueries = flatten(Object.keys(routes).map(routeName => {
    const route = routes[routeName];
    const results = route.context ? getContextGraphQLQueries(route.context, {
      routeName,
      path: route.path,
      isPaginated: false
    }) : [];

    if (route.paginate && route.paginate.data) {
      const filePath = isGraphQLQueryConfig(route.paginate.data) ? route.paginate.data.query : route.paginate.data;
      results.push({
        routeName,
        filePath,
        path: route.path,
        isPaginated: true
      });
    }

    return results;
  }));
  return context ? getContextGraphQLQueries(context, {
    routeName: '*',
    path: '*',
    isPaginated: false
  }).concat(routeQueries) : routeQueries;
}
