// @takeshape/ssg/es/nunjucks.js
import path from 'path';
import TemplateRenderError from './errors/template-render-error';
import codeFilter from './filters/code-filter';
import createImageFilter from './filters/create-image-filter';
import createAssetFilter from './filters/create-asset-filter';
import markdownFilter from './filters/markdown-filter';
import createNumberFilter from './filters/create-number-filter';
import createRouteFilter from './filters/create-route-filter';
import createDateFilter from './filters/create-date-filter';
import pluralizeFilter from './filters/pluralize-filter';
import { intersectionFilter, unionFilter } from './filters/array-filters';
import { permalink } from './paths';
import nunjucks from '@takeshape/vm-nunjucks';
export function fixPrototype(obj) {
  // this is a hack because nunjucks doesn't support objects without a prototype
  return JSON.parse(JSON.stringify(obj));
}

function parseLocation(line) {
  const groups = /\(([^)]+)\)(?: \[Line (\d+)(?:, Column (\d+))?])?/.exec(line);
  return groups ? {
    file: groups[1],
    line: parseInt(groups[2], 10),
    column: parseInt(groups[3], 10)
  } : undefined;
}

function stripMessage(message) {
  return message.replace(/\s*(?:Error:)?\s*/, '');
}

export function formatError(error) {
  const lines = error.message.split('\n');
  const location = parseLocation(lines[0]);
  const message = stripMessage(lines[lines.length - 1]);
  return new TemplateRenderError(message, location);
}

class SyncLoader {
  constructor(fileLoader) {
    this.async = false;
    this.fileLoader = fileLoader;
  }

  // eslint-disable-next-line no-dupe-class-members
  getSource(name) {
    const src = this.fileLoader(name);
    return src === undefined ? undefined : {
      src,
      path: name,
      noCache: false
    };
  }

}

// Manually promisify env.render instead of bluebird.promisify
// vm-nunjucks uses vm2 which disallows access to object.prototype from
// code outside the vm this causes bluebird.promisify to throw a proxy error
function getNunjucksRender(env) {
  return (template, context) => new Promise((resolve, reject) => env.render(template, context, (error, str) => {
    if (error) {
      return reject(error);
    }

    resolve(str || '');
  }));
}

export default function nunjucksFactory(fileLoader, config, stats) {
  const loader = filePath => fileLoader(path.join(config.templatePath, filePath));

  const templateLoader = new SyncLoader(loader);
  const env = new nunjucks.Environment(templateLoader);
  env.addFilter('date', createDateFilter(config));
  env.addFilter('route', createRouteFilter(config, stats));
  env.addFilter('md', markdownFilter);
  const numberFilter = createNumberFilter(config);
  env.addFilter('numberFormat', numberFilter);
  env.addFilter('number', numberFilter);
  env.addFilter('code', codeFilter);
  env.addFilter('image', createImageFilter(config.imageBaseUrl, config.imageDefaults));
  env.addFilter('asset', createAssetFilter(config.assetBaseUrl));
  env.addFilter('intersection', intersectionFilter);
  env.addFilter('union', unionFilter);
  env.addFilter('pluralize', pluralizeFilter);
  const render = getNunjucksRender(env);
  return async (path, template, context) => {
    try {
      stats.pagesGenerated++;
      return {
        path: permalink(path),
        contents: await render(template, fixPrototype(context))
      };
    } catch (e) {
      throw e instanceof TemplateRenderError ? e : formatError(e);
    }
  };
}
