// @takeshape/ssg/lib/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.prepareConfig = prepareConfig;
exports.loadConfig = loadConfig;
exports.localizeConfig = localizeConfig;
exports.isGraphQLQueryConfig = isGraphQLQueryConfig;
exports.getGraphQLQueries = getGraphQLQueries;

var _jsYaml = _interopRequireDefault(require("js-yaml"));

var _jsonVariables = _interopRequireDefault(require("json-variables"));

var _flatten = _interopRequireDefault(require("lodash/flatten"));

var _forEach = _interopRequireDefault(require("lodash/forEach"));

var _merge = _interopRequireDefault(require("lodash/merge"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function prepareConfig(config, envConfig) {
  const combined = (0, _merge.default)({
    env: {},
    context: {},
    routes: {},
    imageBaseUrl: 'https://images.takeshape.io',
    assetBaseUrl: 'https://assets.takeshape.io',
    imageDefaults: {
      auto: 'compress,format'
    },
    locale: 'en-us',
    pathPrefix: '',
    templatePath: 'templates',
    buildPath: 'build',
    staticPath: 'static',
    dates: {
      tz: 'America/New_York',
      format: 'LLL'
    },
    usageStats: false,
    gzip: false
  }, config, envConfig);
  return (0, _jsonVariables.default)(combined, {
    heads: '${',
    tails: '}'
  });
}

async function loadConfig(fileLoader, configFilePath, envConfig) {
  const config = _jsYaml.default.load((await fileLoader(configFilePath || 'tsg.yml')));

  return prepareConfig(config, envConfig);
}

function localizeConfig(config, localeStr) {
  if (!config.locales || !config.locales[localeStr]) {
    throw new Error(`Locale "${localeStr}" is not configured `);
  }

  const locale = config.locales[localeStr];
  return { ...config,
    pathPrefix: locale.pathPrefix,
    locale: localeStr,
    dates: { ...config.dates,
      ...locale.dates
    },
    numbers: { ...config.numbers,
      ...locale.numbers
    }
  };
}

function isGraphQLPath(path) {
  return path.endsWith('.graphql') || path.endsWith('.gql');
}

function isGraphQLQueryConfig(x) {
  return typeof x === 'object' && typeof x.query === 'string' && isGraphQLPath(x.query);
}

function getContextGraphQLQueries(context, info) {
  if (typeof context === 'string' && isGraphQLPath(context)) {
    return [{
      filePath: context,
      ...info
    }];
  }

  if (typeof context === 'object') {
    if (isGraphQLQueryConfig(context)) {
      return [{
        filePath: context.query,
        ...info
      }];
    }

    const result = [];
    (0, _forEach.default)(context, value => {
      Array.prototype.push.apply(result, getContextGraphQLQueries(value, info));
    });
    return result;
  }

  return [];
}

function getGraphQLQueries(config) {
  const {
    context,
    routes
  } = config;
  const routeQueries = (0, _flatten.default)(Object.keys(routes).map(routeName => {
    const route = routes[routeName];
    const results = route.context ? getContextGraphQLQueries(route.context, {
      routeName,
      path: route.path,
      isPaginated: false
    }) : [];

    if (route.paginate && route.paginate.data) {
      const filePath = isGraphQLQueryConfig(route.paginate.data) ? route.paginate.data.query : route.paginate.data;
      results.push({
        routeName,
        filePath,
        path: route.path,
        isPaginated: true
      });
    }

    return results;
  }));
  return context ? getContextGraphQLQueries(context, {
    routeName: '*',
    path: '*',
    isPaginated: false
  }).concat(routeQueries) : routeQueries;
}
