// @takeshape/ssg/lib/nunjucks.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.fixPrototype = fixPrototype;
exports.formatError = formatError;
exports.default = nunjucksFactory;

var _path = _interopRequireDefault(require("path"));

var _templateRenderError = _interopRequireDefault(require("./errors/template-render-error"));

var _codeFilter = _interopRequireDefault(require("./filters/code-filter"));

var _createImageFilter = _interopRequireDefault(require("./filters/create-image-filter"));

var _createAssetFilter = _interopRequireDefault(require("./filters/create-asset-filter"));

var _markdownFilter = _interopRequireDefault(require("./filters/markdown-filter"));

var _createNumberFilter = _interopRequireDefault(require("./filters/create-number-filter"));

var _createRouteFilter = _interopRequireDefault(require("./filters/create-route-filter"));

var _createDateFilter = _interopRequireDefault(require("./filters/create-date-filter"));

var _pluralizeFilter = _interopRequireDefault(require("./filters/pluralize-filter"));

var _arrayFilters = require("./filters/array-filters");

var _paths = require("./paths");

var _vmNunjucks = _interopRequireDefault(require("@takeshape/vm-nunjucks"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function fixPrototype(obj) {
  // this is a hack because nunjucks doesn't support objects without a prototype
  return JSON.parse(JSON.stringify(obj));
}

function parseLocation(line) {
  const groups = /\(([^)]+)\)(?: \[Line (\d+)(?:, Column (\d+))?])?/.exec(line);
  return groups ? {
    file: groups[1],
    line: parseInt(groups[2], 10),
    column: parseInt(groups[3], 10)
  } : undefined;
}

function stripMessage(message) {
  return message.replace(/\s*(?:Error:)?\s*/, '');
}

function formatError(error) {
  const lines = error.message.split('\n');
  const location = parseLocation(lines[0]);
  const message = stripMessage(lines[lines.length - 1]);
  return new _templateRenderError.default(message, location);
}

class SyncLoader {
  constructor(fileLoader) {
    this.async = false;
    this.fileLoader = fileLoader;
  }

  // eslint-disable-next-line no-dupe-class-members
  getSource(name) {
    const src = this.fileLoader(name);
    return src === undefined ? undefined : {
      src,
      path: name,
      noCache: false
    };
  }

}

// Manually promisify env.render instead of bluebird.promisify
// vm-nunjucks uses vm2 which disallows access to object.prototype from
// code outside the vm this causes bluebird.promisify to throw a proxy error
function getNunjucksRender(env) {
  return (template, context) => new Promise((resolve, reject) => env.render(template, context, (error, str) => {
    if (error) {
      return reject(error);
    }

    resolve(str || '');
  }));
}

function nunjucksFactory(fileLoader, config, stats) {
  const loader = filePath => fileLoader(_path.default.join(config.templatePath, filePath));

  const templateLoader = new SyncLoader(loader);
  const env = new _vmNunjucks.default.Environment(templateLoader);
  env.addFilter('date', (0, _createDateFilter.default)(config));
  env.addFilter('route', (0, _createRouteFilter.default)(config, stats));
  env.addFilter('md', _markdownFilter.default);
  const numberFilter = (0, _createNumberFilter.default)(config);
  env.addFilter('numberFormat', numberFilter);
  env.addFilter('number', numberFilter);
  env.addFilter('code', _codeFilter.default);
  env.addFilter('image', (0, _createImageFilter.default)(config.imageBaseUrl, config.imageDefaults));
  env.addFilter('asset', (0, _createAssetFilter.default)(config.assetBaseUrl));
  env.addFilter('intersection', _arrayFilters.intersectionFilter);
  env.addFilter('union', _arrayFilters.unionFilter);
  env.addFilter('pluralize', _pluralizeFilter.default);
  const render = getNunjucksRender(env);
  return async (path, template, context) => {
    try {
      stats.pagesGenerated++;
      return {
        path: (0, _paths.permalink)(path),
        contents: await render(template, fixPrototype(context))
      };
    } catch (e) {
      throw e instanceof _templateRenderError.default ? e : formatError(e);
    }
  };
}
