// @takeshape/ssg/lib/resolve-context.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getType = getType;
exports.handleJSON = handleJSON;
exports.handleYaml = handleYaml;
exports.handleGraphQL = handleGraphQL;
exports.handleContextString = handleContextString;
exports.resolveGraphQLConfig = resolveGraphQLConfig;
exports.default = resolveContext;

var _bluebird = _interopRequireDefault(require("bluebird"));

var _forEach = _interopRequireDefault(require("lodash/forEach"));

var _jsYaml = _interopRequireDefault(require("js-yaml"));

var _path = _interopRequireDefault(require("path"));

var _graphqlError = _interopRequireDefault(require("./errors/graphql-error"));

var _stats = require("./stats");

var _config = require("./config");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getType(path) {
  const matches = /\.(graphql|yml|yaml|json)$/.exec(path);

  if (matches && matches.length) {
    const type = matches[1];
    return type === 'yml' ? 'yaml' : type;
  }

  return null;
}

function getPath(contextConfig) {
  if (typeof contextConfig === 'string') {
    return contextConfig;
  }

  if ((0, _config.isGraphQLQueryConfig)(contextConfig)) {
    return contextConfig.query;
  }

  throw new Error('Unable to find file path in context config');
}

function handleJSON({
  srcPath,
  filePath,
  fileLoader
}) {
  return fileLoader(_path.default.join(srcPath, filePath)).then(JSON.parse);
}

function handleYaml({
  srcPath,
  filePath,
  fileLoader
}) {
  return fileLoader(_path.default.join(srcPath, filePath)).then(_jsYaml.default.safeLoad);
}

async function handleGraphQL({
  srcPath,
  filePath,
  variables,
  connector,
  fileLoader,
  route,
  stats
}) {
  const queryPath = _path.default.join(srcPath, filePath);

  const query = await fileLoader(queryPath);
  const res = await connector({
    query,
    variables
  }, {
    stats,
    source: filePath
  });

  if (res.errors) {
    throw new _graphqlError.default(`An error occurred while resolving ${queryPath}`, res.errors);
  }

  const data = res.data;

  if (route && route.path && !route.paginate) {
    (0, _stats.recordContentUsage)(stats, data, route.path);
  }

  return data;
}

async function handleContextString(params) {
  const {
    contextConfig,
    ...rest
  } = params;
  const filePath = getPath(contextConfig);
  const type = getType(filePath);

  if (type === 'json') {
    return handleJSON({ ...rest,
      filePath
    });
  }

  if (type === 'yaml') {
    return handleYaml({ ...rest,
      filePath
    });
  }

  if (type === 'graphql') {
    return handleGraphQL({ ...rest,
      filePath
    });
  }

  return filePath;
}

async function resolveGraphQLConfig(config, {
  fileLoader,
  srcPath
}) {
  let source;
  let variables;

  if (typeof config === 'string') {
    source = config;
  } else if ((0, _config.isGraphQLQueryConfig)(config)) {
    source = config.query;
    variables = config.variables;
  }

  if (!source) {
    throw new Error('Missing .graphql or .gql query file');
  }

  const filePath = _path.default.join(srcPath, source);

  const query = await fileLoader(filePath);
  return {
    source,
    query,
    variables
  };
}

function handleValue(contextParams) {
  const {
    contextConfig,
    ...rest
  } = contextParams;

  if (typeof contextConfig === 'string') {
    return handleContextString(contextParams);
  }

  if (typeof contextConfig === 'object') {
    if ((0, _config.isGraphQLQueryConfig)(contextConfig)) {
      return handleGraphQL({ ...rest,
        filePath: contextConfig.query,
        variables: contextConfig.variables
      });
    }

    const context = {};
    (0, _forEach.default)(contextConfig, (value, key) => {
      context[key] = handleValue({ ...rest,
        contextConfig: value
      });
    });
    return _bluebird.default.props(context);
  }

  return contextConfig;
}

function isObject(obj) {
  return Boolean(obj) && typeof obj === 'object' && !Array.isArray(obj);
}

async function resolveContext(contextParams) {
  const context = await handleValue(contextParams);
  return isObject(context) ? context : {};
}
