// @takeshape/cli/lib/commands/export.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _fsExtra = require("fs-extra");

var _nodeFetch = _interopRequireDefault(require("node-fetch"));

var _chalk = _interopRequireDefault(require("chalk"));

var _streams = require("@takeshape/streams");

var _spin = _interopRequireDefault(require("../util/spin"));

var _log = _interopRequireDefault(require("../log"));

var _graphql = require("../graphql");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const startExport = (0, _graphql.graphQLQuery)(`
  mutation ($empty: Boolean) {
    result: tsExportProject(empty: $empty, origin: "@takeshape/cli")
  }
`);
const getExport = (0, _graphql.graphQLQuery)(`
  query ($id: String!) {
    result: tsGetProjectExport(id: $id) {
      id
      status
      file
    }
  }
`);

async function downloadExport(projectExport, path) {
  // Download the file to the current directory
  const dirName = `export-${projectExport.id}`;
  const res = await (0, _nodeFetch.default)(projectExport.file);
  await (0, _streams.pump)(res.body, (0, _fsExtra.createWriteStream)(`${path}/${dirName}.zip`));
  return dirName;
}
/*
tsg export

Create and download a project export

Options:
  --without-data: download an empty project, defaults to false
  --output: the path to save the download, defaults to current directory
*/


var _default = async (command, params, flags) => {
  const authToken = params.authToken;
  const projectId = params.projectId;
  const withoutData = flags.withoutData || false;
  const path = flags.output || '.';

  if (!projectId) {
    (0, _log.default)(_chalk.default.red('Error: No project ID provided. Please run `tsg init`.'));
    process.exit(1);
  }

  if (!authToken) {
    (0, _log.default)(_chalk.default.red('Error: No auth token provided. Please run `tsg init`.'));
    process.exit(1);
  }

  const client = (0, _graphql.createConnector)(params, authToken, projectId);

  try {
    const exportId = await startExport(client, {
      empty: withoutData
    });
    (0, _log.default)(`Exporting project…\n`);
    const spinQuery = getExport.bind(getExport, client, {
      id: exportId
    });
    await (0, _spin.default)(spinQuery);
    const projectExport = await getExport(client, {
      id: exportId
    });
    (0, _log.default)('Downloading the export…\n');
    const dirName = await downloadExport(projectExport, path);
    (0, _log.default)(`Success! Downloaded to ${path}/${dirName}.zip\n`);
  } catch (error) {
    (0, _log.default)(_chalk.default.red('Error: ', error.message));
    process.exit(1);
  }
};

exports.default = _default;
