// @takeshape/ssg/es/graphql/analyze.js
import BbPromise from 'bluebird';
import { TypeInfo, visit, visitWithTypeInfo, isObjectType, isListType, getLocation } from 'graphql';
import { parse } from './ast';
import { getFields, getItemType } from './query';
import { getGraphQLQueries } from '../config';
import path from 'path';
const startLocation = {
  line: 0,
  column: 0
};
export function getFieldUsage(queryAst, schema, fileInfo) {
  const typeInfo = new TypeInfo(schema);
  const result = {};
  const visitor = {
    Field: {
      enter(node) {
        const type = getItemType(typeInfo.getFieldDef().type);
        const fields = getFields(type);
        const fieldInfo = result[type.toString()] || {};

        if (fields && node.selectionSet) {
          node.selectionSet.selections.forEach(selection => {
            if (selection.kind === 'Field' || selection.kind === 'FragmentSpread') {
              const name = selection.name.value;
              const usage = fieldInfo[name] ? fieldInfo[name].usage : [];
              const source = {
                body: fileInfo.body,
                name: fileInfo.name,
                locationOffset: startLocation
              };
              fieldInfo[name] = {
                type: fields[name].type.toString(),
                usage: usage.concat({
                  file: fileInfo.name,
                  location: selection.loc ? getLocation(source, selection.loc.start) : startLocation
                })
              };
            }
          });
          result[type.toString()] = fieldInfo;
        }
      }

    }
  };
  visit(queryAst, visitWithTypeInfo(typeInfo, visitor));
  return result;
}
export function detectTypes(queryAst, schema) {
  const typeInfo = new TypeInfo(schema);
  const result = new Set();
  const queryType = schema.getQueryType();

  if (!queryType) {
    throw new Error('Invalid schema has no queries');
  }

  const topLevelQueries = new Set(Object.keys(queryType.getFields()));
  const visitor = {
    Field: {
      enter(node) {
        if (topLevelQueries.has(node.name.value)) {
          const fieldType = typeInfo.getFieldDef().type;

          if (isObjectType(fieldType)) {
            const fields = fieldType.getFields();

            if (fields.items && isListType(fields.items.type)) {
              result.add(fields.items.type.ofType.toString());
            } else {
              result.add(fieldType.toString());
            }
          }
        }
      }

    }
  };
  visit(queryAst, visitWithTypeInfo(typeInfo, visitor));
  return result;
}
export async function getTypeUsage({
  config,
  schema,
  fileLoader
}) {
  const queries = getGraphQLQueries(config);
  return BbPromise.reduce(queries, async (result, routeQuery) => {
    const {
      filePath,
      ...info
    } = routeQuery;
    const queryString = await fileLoader(path.join(config.templatePath, filePath));
    const types = detectTypes(parse(queryString), schema);

    for (const type of types) {
      if (!result[type]) result[type] = [];
      result[type].push(info);
    }

    return result;
  }, {});
}
