// @takeshape/ssg/es/graphql/ast.js
import { parse as baseParse } from 'graphql';
export function parse(source, options) {
  return baseParse(source, options);
}
export function createArgument({
  name,
  value
}) {
  return {
    kind: 'Argument',
    name: {
      kind: 'Name',
      value: name
    },
    value: {
      kind: 'Variable',
      name: {
        kind: 'Name',
        value
      }
    }
  };
}

function notNull(type) {
  return {
    kind: 'NonNullType',
    type
  };
}

function createType(typeString) {
  const isNotNull = typeString.endsWith('!');
  const value = isNotNull ? typeString.slice(0, typeString.length - 1) : typeString;
  const namedType = {
    kind: 'NamedType',
    name: {
      kind: 'Name',
      value
    }
  };
  return isNotNull ? notNull(namedType) : namedType;
}

export function createVariableDefinition({
  name,
  type
}) {
  return {
    kind: 'VariableDefinition',
    directives: [],
    variable: {
      kind: 'Variable',
      name: {
        kind: 'Name',
        value: name
      }
    },
    type: createType(type)
  };
}
export function createSelectionSet(selections) {
  return {
    kind: 'SelectionSet',
    selections
  };
}
export function createField(name) {
  return {
    kind: 'Field',
    name: {
      kind: 'Name',
      value: name
    },
    arguments: [],
    directives: []
  };
}
export function isQueryNode(node) {
  return node.kind === 'OperationDefinition' && node.operation === 'query';
}
export function findQuery(ast) {
  return ast.definitions.find(isQueryNode);
}
export function addQueryVariables(variables, ast) {
  const query = findQuery(ast);

  if (query) {
    const variableNames = new Set(variables.map(variable => variable.name));

    if (query.variableDefinitions) {
      query.variableDefinitions = query.variableDefinitions.filter(def => !variableNames.has(def.variable.name.value)).concat(variables.map(createVariableDefinition));
    }
  }

  return ast;
}
export function wrapSelectionSet(selectionSet, name, args = []) {
  return {
    kind: 'SelectionSet',
    selections: [{
      kind: 'Field',
      name: {
        kind: 'Name',
        value: name
      },
      arguments: args,
      directives: [],
      selectionSet
    }]
  };
}
export function wrapQuery(name, args, ast) {
  const query = findQuery(ast);

  if (query) {
    const argsAst = args.map(createArgument);
    query.selectionSet = wrapSelectionSet(query.selectionSet, name, argsAst);
  }

  return ast;
}
