// @takeshape/ssg/es/graphql/query.js
import { print, TypeInfo, visit, visitWithTypeInfo } from 'graphql';
import { addQueryVariables, wrapQuery, parse, createArgument, findQuery, createVariableDefinition } from './ast';
export function getItemType(fieldType) {
  return 'ofType' in fieldType ? fieldType.ofType : fieldType;
}
export function getFields(type) {
  return 'getFields' in type ? type.getFields() : undefined;
}

function addId(node) {
  const selections = node.selectionSet && node.selectionSet.selections;

  if (selections && !selections.find(selection => (selection.kind === 'Field' || selection.kind === 'FragmentSpread') && selection.name.value === '_id')) {
    selections.push({
      kind: 'Field',
      name: {
        kind: 'Name',
        value: '_id'
      },
      arguments: [],
      directives: []
    });
    return node;
  }
}

export function ensureIds(queryAst, schema) {
  const typeInfo = new TypeInfo(schema);
  const visitor = {
    Field: {
      enter(node) {
        const fieldDef = typeInfo.getFieldDef();

        if (fieldDef) {
          const type = getItemType(fieldDef.type);
          const fields = getFields(type);

          if (fields && fields._id) {
            return addId(node);
          }
        }
      }

    }
  };
  return visit(queryAst, visitWithTypeInfo(typeInfo, visitor));
}
const addLocaleVariable = addQueryVariables.bind(null, [{
  name: 'locale',
  type: 'String'
}]);
const wrapWithContext = wrapQuery.bind(null, 'withContext', [{
  name: 'locale',
  value: 'locale'
}]);
export function wrapWithLocaleContext(ast) {
  return wrapWithContext(addLocaleVariable(ast));
}

function unboxLocalizedResponse(res) {
  return { ...res,
    data: res.data && res.data.withContext
  };
}

export const FILTER_VAR_NAME = '__filter';
export function addContentIdFilter(ast, schema) {
  const typeInfo = new TypeInfo(schema);
  let filterType;
  const varsToRemove = new Set();
  const visitor = {
    Field: {
      enter(node) {
        const matches = node.name.value.match(/^get(.+)List$/);

        if (matches) {
          const fieldDef = typeInfo.getFieldDef();
          const filterArgType = fieldDef.args.find(arg => arg.name === 'filter');

          if (filterArgType) {
            filterType = filterArgType.type.toString();
            const filterArgAst = createArgument({
              name: 'filter',
              value: FILTER_VAR_NAME
            });
            return { ...node,
              arguments: node.arguments ? node.arguments.filter(arg => {
                if (arg.name.value === 'filter' && arg.value.kind === 'Variable') {
                  varsToRemove.add(arg.value.name.value);
                  return false;
                }

                return true;
              }).concat(filterArgAst) : [filterArgAst]
            };
          }
        }
      }

    }
  };
  const updatedAst = visit(ast, visitWithTypeInfo(typeInfo, visitor));

  if (filterType) {
    const query = findQuery(updatedAst);

    if (query) {
      const variableDef = createVariableDefinition({
        name: FILTER_VAR_NAME,
        type: filterType
      });
      query.variableDefinitions = query.variableDefinitions ? query.variableDefinitions.filter(def => !varsToRemove.has(def.variable.name.value)).concat(variableDef) : [variableDef];
    }
  }

  return {
    ast: updatedAst,
    varName: FILTER_VAR_NAME,
    modified: Boolean(filterType)
  };
}
export function withQueryTransforms(connector, schema, {
  locale,
  usage,
  contentId
} = {}) {
  return async (gqlParams, context) => {
    const variables = { ...gqlParams.variables
    };
    let queryAst = parse(gqlParams.query);

    if (usage) {
      queryAst = ensureIds(queryAst, schema);
    }

    if (context && context.isPaginated && contentId) {
      const {
        ast,
        modified,
        varName
      } = addContentIdFilter(queryAst, schema);

      if (modified) {
        queryAst = ast;
        variables[varName] = {
          term: {
            _id: contentId
          }
        };
      }
    }

    if (locale) {
      variables.locale = variables.locale || locale;
      queryAst = wrapWithLocaleContext(queryAst);
    }

    const res = await connector({
      query: print(queryAst),
      variables
    });
    return locale ? unboxLocalizedResponse(res) : res;
  };
}
