// @takeshape/ssg/es/graphql/schema-connector-factory.js
import fse from 'fs-extra';
import fetch from 'node-fetch';
import path from 'path';
import { graphql, formatError } from 'graphql';
import { connectorWithMigrations } from './migrate';
export function getRemoteConnector(config) {
  const {
    url
  } = config;

  if (!url) {
    throw new Error('Invalid GraphQL config: missing url');
  }

  return params => {
    return fetch(url, {
      method: 'POST',
      body: JSON.stringify(params),
      headers: {
        'Content-Type': 'application/json',
        ...(config.headers || {})
      }
    }).then(res => {
      if (res.ok) {
        return res.json();
      }

      return res.text().then(text => {
        throw new Error(`${res.statusText}: ${text}`);
      });
    });
  };
}

function parseConfigJson(json) {
  if (json.request) {
    return {
      type: 'request',
      ...json.request
    };
  }

  if (json['graphql-js']) {
    return {
      type: 'graphql-js',
      file: json['graphql-js']
    };
  }

  return null;
}

function readConfigFile(path) {
  let json;

  try {
    json = JSON.parse(fse.readFileSync(`${path}/.graphqlrc`, 'utf-8'));
  } catch (e) {
    throw new Error(`An error occurred loading .graphqlrc.`);
  }

  const config = parseConfigJson(json);

  if (config) {
    return config;
  }

  throw new Error(`Invalid configuration file: ${JSON.stringify(json)}`);
}

export function schemaConnectorFactory(cwd, config) {
  config = config || readConfigFile(cwd);
  let connector;

  if (config.type === 'graphql-js' && config.file) {
    const schema = require(path.join(cwd, config.file)); // eslint-disable-line


    connector = async params => {
      const res = await graphql(schema.default || schema, params.query, null, {}, params.variables);
      return {
        data: res.data,
        errors: res.errors && res.errors.map(formatError)
      };
    };
  }

  if (config.type === 'request') {
    connector = getRemoteConnector(config);
  }

  if (connector) {
    return connectorWithMigrations(connector);
  }

  throw new Error('Could not create GraphQL connector - invalid configuration');
}
