// @takeshape/cli/lib/util/api.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isJWT = isJWT;
exports.getAuthHeader = getAuthHeader;
exports.default = api;

var _nodeFetch = _interopRequireDefault(require("node-fetch"));

var _errors = require("../errors");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function isValidHeader(str) {
  const json = Buffer.from(str, 'base64').toString('utf8');

  try {
    const header = JSON.parse(json);
    return header.typ === 'JWT';
  } catch (e) {
    return false;
  }
}

function isJWT(str) {
  const parts = str.split('.');
  return parts.length === 3 && isValidHeader(parts[0]);
}

function getAuthHeader(authToken) {
  if (!authToken) {
    return {};
  }

  return isJWT(authToken) ? {
    'X-TakeShape-Token': authToken
  } : {
    Authorization: `Bearer ${authToken}`
  };
}

async function api(params, method, path, body) {
  const requestParams = {
    method,
    headers: { ...getAuthHeader(params.authToken),
      'Content-Type': 'application/json'
    }
  };

  if (body) {
    requestParams.body = JSON.stringify(body);
  }

  const endpoint = `${params.endpoint}${path}`;
  const res = await (0, _nodeFetch.default)(endpoint, requestParams);

  if (res.ok) {
    return res.json();
  }

  throw new _errors.HTTPError(`${res.statusText}`, res.status);
}
