// @takeshape/cli/lib/util/pusher.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.handleWarning = handleWarning;
exports.handleAction = handleAction;
exports.subscribe = subscribe;

var _api = _interopRequireDefault(require("./api"));

var _pusherClient = _interopRequireDefault(require("pusher-client"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function handleWarning(type, error) {
  if (type === 'Error' && error.type === 'WebSocketError') {
    console.log('Disconnected from TakeShape. Restart this process to resume live content updates.');
  }
}

_pusherClient.default.warn = handleWarning;

function handleAction(callback) {
  return action => {
    if (action.type === 'content/CONTENT_UPDATED' || action.type === 'content/CONTENT_CREATED' || action.type === 'content/CONTENT_DELETED') {
      callback(action);
    }
  };
}

async function subscribe(params, callback) {
  const config = await (0, _api.default)(params, 'GET', `/project/${params.projectId}/pusher-client-config`);
  const pusher = new _pusherClient.default(config.key, {
    authEndpoint: `${params.endpoint}/project/channel-auth?auth=${params.authToken}`,
    cluster: config.cluster,
    encrypted: true
  });
  const channel = pusher.subscribe(`presence-project.${params.projectId}`);
  channel.bind('server-action', handleAction(callback));
}
