// @takeshape/cli/lib/util/upload.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createTempFile = createTempFile;
exports.getUploadUrl = getUploadUrl;
exports.uploadFile = uploadFile;

var _api = _interopRequireDefault(require("./api"));

var _tmp = _interopRequireDefault(require("tmp"));

var _path = _interopRequireDefault(require("path"));

var _fsExtra = _interopRequireDefault(require("fs-extra"));

var _request = _interopRequireDefault(require("request"));

var _streams = require("@takeshape/streams");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function createTempFile(filePath) {
  const tmpDir = _tmp.default.dirSync({
    unsafeCleanup: true
  });

  return [_path.default.join(tmpDir.name, filePath), tmpDir.removeCallback];
}

async function getUploadUrl(params) {
  try {
    const {
      uploadUrl
    } = await (0, _api.default)(params, 'GET', `/project/${params.projectId}/ssg/${params.siteId}/upload-url`);
    return uploadUrl;
  } catch (e) {
    throw new Error(`Failed to get upload url: ${e.message}`);
  }
}

async function uploadFile(filePath, uploadUrl) {
  const stats = _fsExtra.default.statSync(filePath);

  const putStream = (0, _request.default)({
    method: 'PUT',
    url: uploadUrl,
    headers: {
      'Content-Length': stats.size
    }
  });
  await (0, _streams.pump)(_fsExtra.default.createReadStream(filePath), putStream);
  return stats.size;
}
