// @takeshape/cli/lib/commands/init/login.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.loginBrowser = loginBrowser;
exports.attemptLogin = attemptLogin;
exports.attemptOpenForgotPassword = attemptOpenForgotPassword;
exports.loginCli = loginCli;
exports.loginQuestions = void 0;

var _http = _interopRequireDefault(require("http"));

var _open = _interopRequireDefault(require("open"));

var _prompt = require("../../prompt");

var _emailValidator = require("email-validator");

var _getPort = _interopRequireDefault(require("get-port"));

var _saga = require("../../saga");

var _auth = require("../../auth");

var _messages = require("./messages");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

async function loginBrowser(params) {
  // Port for login page to communicate with the CLI
  const port = await (0, _getPort.default)(); // Open a browser to the login page

  await (0, _open.default)(`${params.appUrl}/login?returnToCli=true&cliPort=${port}`, {
    wait: false
  }); // Spin up an ad-hoc server to recive the auth token from the login page

  return new Promise((resolve, reject) => {
    const server = _http.default.createServer((req, res) => {
      res.setHeader('Content-Type', 'text/plain');
      res.setHeader('Access-Control-Allow-Origin', params.appUrl);
      res.setHeader('Access-Control-Allow-Headers', 'x-takeshape-token');

      if (req.method === 'OPTIONS') {
        res.statusCode = 200;
        res.end('Success');
        return;
      }

      const token = req.headers['x-takeshape-token'];

      if (token) {
        res.statusCode = 200;
        res.end('Success');
        (0, _messages.loginSuccess)();
        resolve(token);
      } else {
        res.statusCode = 400;
        res.end('Bad Request');
        reject(new Error('Did not receive auth token from webapp.'));
      }

      server.close();
    });

    server.listen(port, '127.0.0.1', () => {
      (0, _messages.waitingForLogin)();
    });
  });
}

const loginQuestions = [{
  type: 'input',
  name: 'email',
  message: 'Enter your TakeShape email:',

  validate(input) {
    return (0, _emailValidator.validate)(input) || 'Please enter a valid email';
  }

}, {
  type: 'password',
  name: 'password',
  message: 'Enter your TakeShape password:'
}];
exports.loginQuestions = loginQuestions;

function* attemptLogin(params) {
  const answers = yield (0, _saga.call)(_prompt.prompt, loginQuestions);
  return yield (0, _saga.call)(_auth.login, params.endpoint, answers.email, answers.password);
}

function* attemptOpenForgotPassword(params) {
  const isConfirmed = yield (0, _saga.call)(_prompt.confirm, 'Would you like to open TakeShape to set a password?');

  if (isConfirmed) {
    yield (0, _saga.call)(_open.default, params.appUrl + '/forgot-password', {
      wait: false
    });
  }
}

function* loginCli(params, maxAttempts) {
  let attempts = 0;

  do {
    attempts++;

    try {
      return yield (0, _saga.call)(attemptLogin, params);
    } catch (e) {
      if (e.statusCode === 403) {
        attempts--;
        yield (0, _saga.call)(_messages.missingPasswordMsg);
        yield (0, _saga.call)(attemptOpenForgotPassword, params);
      } else {
        yield (0, _saga.call)(_messages.invalidLoginMsg);
      }
    }
  } while (attempts < maxAttempts);

  yield (0, _saga.call)(_messages.forgotPasswordWarning, params);
}
