// @takeshape/ssg/es/generate/context.js
import { formatPath } from '@takeshape/routing';
export function combineContext(env, globalContext, route, routeContext) {
  return { ...globalContext,
    ...routeContext,
    global: globalContext,
    env,
    environmentVariables: env,
    // DEPRECATE NEXT MAJOR VERSION
    currentPath: route.path
  };
}
export const getListPageContext = (items, pageNumber, routeContext) => {
  const {
    route,
    context,
    totalPages
  } = routeContext;
  const {
    paginate
  } = route;
  let currentPath;

  if (paginate.firstPage && pageNumber === 1) {
    currentPath = paginate.firstPage;
  } else {
    currentPath = formatPath(paginate.path, {
      num: pageNumber
    }).path;
  }

  const pageContext = {
    currentPath,
    items,
    pagination: {
      currentPage: pageNumber,
      totalPages
    }
  };
  return { ...context,
    ...pageContext
  };
};
export const getItemContext = (route, context, {
  current,
  previous,
  next
}) => {
  const itemContext = {
    item: current.item,
    currentPath: current.path
  };

  if (route.paginate && route.paginate.itemName) {
    itemContext[route.paginate.itemName] = current.item;
  }

  if (previous) {
    itemContext.previousItem = previous.item;
    itemContext.previousPath = previous.path;
  }

  if (next) {
    itemContext.nextItem = next.item;
    itemContext.nextPath = next.path;
  }

  return { ...context,
    ...itemContext
  };
};
