// @takeshape/cli/lib/commands/deploy/zip.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createZip = createZip;
exports.deployZip = deployZip;

var _path = _interopRequireDefault(require("path"));

var _fsExtra = _interopRequireDefault(require("fs-extra"));

var _archiver = _interopRequireDefault(require("archiver"));

var _streamToPromise = _interopRequireDefault(require("stream-to-promise"));

var _upload = require("../../util/upload");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function createZip({
  projectPath,
  staticPath,
  templatePath,
  configFilePath,
  zipPath
}) {
  const output = _fsExtra.default.createWriteStream(zipPath);

  const archive = (0, _archiver.default)('zip');
  archive.pipe(output);
  archive.file(configFilePath, {
    name: 'tsg.yml'
  });
  archive.directory(_path.default.join(projectPath, staticPath), staticPath);
  archive.directory(_path.default.join(projectPath, templatePath), templatePath);
  archive.finalize();
  return (0, _streamToPromise.default)(output);
}

async function deployZip(params) {
  const [zipPath, removeTempFile] = (0, _upload.createTempFile)('deploy.zip');

  try {
    const [uploadUrl] = await Promise.all([(0, _upload.getUploadUrl)(params.params), createZip({
      zipPath,
      configFilePath: params.params.configFilePath,
      projectPath: params.projectPath,
      templatePath: params.config.templatePath,
      staticPath: params.config.staticPath
    })]);
    const zipSize = await (0, _upload.uploadFile)(zipPath, uploadUrl);
    return zipSize;
  } finally {
    removeTempFile();
  }
}
